### ISSUE TITLE
Release of jenkins 2.0 seems to break this role

### ISSUE THREAD

**Issue by blaffoy** [geerlingguy/ansible-role-jenkins#47]
With the release of [Jenkins 2.0](http://mirrors.jenkins-ci.org/debian/jenkins_2.0_all.deb), the deployment procedure has changed somewhat. On first installing and starting, Jenkins 2.0 wants to lead you through a one time set up of an admin user. In order to do so, you must enter a one time password written to /

This password should be pasted into the web UI in order to allow configuration of an admin user.

Until such a user is set up, Jenkins will deny any attempts to use jenkins-cli.jar. This role relies on that CLI interface to install plugins (and I would guess for other features too).

I think that being able to specify which version of Jenkins you wish to install would workaround this problem, as in this issue: https://github.com/geerlingguy/ansible-role-jenkins/issues/38

**Comment by blaffoy** [comment]
Some output from an attempt to deploy:

```
failed: [jenkins] => (item=xcode-plugin) => {"changed": true, "cmd": ["java", "-jar", "/opt/jenkins-cli.jar", "-s", "http://localhost:8080/", "install-plugin", "xcode-plugin"], "delta": "0:00:00.140143", "end": "2016-04-21 10:27:19.626647", "failed": true, "item": "xcode-plugin", "rc": 255, "start": "2016-04-21 10:27:19.486504", "stderr": "java.io.IOException: No X-Jenkins-CLI2-Port among [null, X-Required-Permission, X-Jenkins, X-You-Are-In-Group, X-Hudson, Content-Length, Expires, X-You-Are-Authenticated-As, X-Permission-Implied-By, Set-Cookie, Server, X-Content-Type-Options, Date, X-Jenkins-Session, Content-Type]\n\tat hudson.cli.CLI.getCliTcpPort(CLI.java:290)\n\tat hudson.cli.CLI.<init>(CLI.java:128)\n\tat hudson.cli.CLIConnectionFactory.connect(CLIConnectionFactory.java:72)\n\tat hudson.cli.CLI._main(CLI.java:479)\n\tat hudson.cli.CLI.main(CLI.java:390)\n\tSuppressed: java.io.IOException: Server returned HTTP response code: 403 for URL: http://localhost:8080/cli\n\t\tat sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1628)\n\t\tat hudson.cli.FullDuplexHttpStream.<init>(FullDuplexHttpStream.java:78)\n\t\tat hudson.cli.CLI.connectViaHttp(CLI.java:158)\n\t\tat hudson.cli.CLI.<init>(CLI.java:132)\n\t\t... 3 more", "stdout": "", "stdout_lines": [], "warnings": []}
failed: [jenkins] => (item=xunit) => {"changed": true, "cmd": ["java", "-jar", "/opt/jenkins-cli.jar", "-s", "http://localhost:8080/", "install-plugin", "xunit"], "delta": "0:00:00.121791", "end": "2016-04-21 10:27:19.852277", "failed": true, "item": "xunit", "rc": 255, "start": "2016-04-21 10:27:19.730486", "stderr": "java.io.IOException: No X-Jenkins-CLI2-Port among [null, X-Required-Permission, X-Jenkins, X-You-Are-In-Group, X-Hudson, Content-Length, Expires, X-You-Are-Authenticated-As, X-Permission-Implied-By, Set-Cookie, Server, X-Content-Type-Options, Date, X-Jenkins-Session, Content-Type]\n\tat hudson.cli.CLI.getCliTcpPort(CLI.java:290)\n\tat hudson.cli.CLI.<init>(CLI.java:128)\n\tat hudson.cli.CLIConnectionFactory.connect(CLIConnectionFactory.java:72)\n\tat hudson.cli.CLI._main(CLI.java:479)\n\tat hudson.cli.CLI.main(CLI.java:390)\n\tSuppressed: java.io.IOException: Server returned HTTP response code: 403 for URL: http://localhost:8080/cli\n\t\tat sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1628)\n\t\tat hudson.cli.FullDuplexHttpStream.<init>(FullDuplexHttpStream.java:78)\n\t\tat hudson.cli.CLI.connectViaHttp(CLI.java:158)\n\t\tat hudson.cli.CLI.<init>(CLI.java:132)\n\t\t... 3 more", "stdout": "", "stdout_lines": [], "warnings": []}

```

**Comment by blaffoy** [comment]
Right now, you can work around this issue by over-riding the repo var to look for the LTS release of Jenkins

```
# For RedHat/CentOS LTS:
jenkins_repo_url: http://pkg.jenkins-ci.org/redhat-stable/jenkins.repo
jenkins_repo_key_url: http://pkg.jenkins-ci.org/redhat-stable/jenkins-ci.org.key
# For Debian/Ubuntu LTS:
jenkins_repo_url: deb http://pkg.jenkins-ci.org/debian-stable binary/
jenkins_repo_key_url: http://pkg.jenkins-ci.org/debian-stable/jenkins-ci.org.key
```

I think it's safe to say that this work around won't work forever.

**Comment by noidi** [comment]
Here's a workaround that should work a bit more reliably:

    pre_tasks:
      # Work around https://github.com/geerlingguy/ansible-role-jenkins/issues/47
      - name: Fetch Jenkins 1.658
        get_url:
          url: http://pkg.jenkins-ci.org/debian/binary/jenkins_1.658_all.deb
          dest: /tmp/jenkins_1.658_all.deb

      - name: Install Jenkins 1.658
        apt: deb=/tmp/jenkins_1.658_all.deb

**Comment by geerlingguy** [comment]
I haven't had time to test with 2.0 yet (though I would like to do so very soon—pipelines/Jenkinsfiles are very enticing), but I'll try to make the role work with 2.0 soon.

**Comment by geerlingguy** [comment]
See also: https://github.com/geerlingguy/ansible-role-jenkins/pull/48

**Comment by ronsonmaria** [comment]
Thanks to blaffoy, we add:

jenkins_repo_url: deb http://pkg.jenkins-ci.org/debian-stable binary/
jenkins_repo_key_url: http://pkg.jenkins-ci.org/debian-stable/jenkins-ci.org.key

in config.yml and work fine!

thanks!

**Comment by blaffoy** [comment]
@ronsonmaria actually, I would recommend using @noidi's workaround to pre-fetch and install a specific version of Jenkins that he laid out code for above. I sketched an alternative that was friendly to `yum` based systems:

```
  pre_tasks:
    # Work around https://github.com/geerlingguy/ansible-role-jenkins/issues/47
    - name: Fetch Jenkins 1.651
      get_url:
        url: http://mirrors.jenkins-ci.org/redhat-stable/jenkins-1.651.1-1.1.noarch.rpm
        dest: /tmp/jenkins-1.651.1-1.1.noarch.rpm
    - name: Install Jenkins 1.651
      yum: name=/tmp/jenkins-1.651.1-1.1.noarch.rpm
```

My workaround will probably break as soon as Jenkins 2.0 is promoted to "stable"

**Comment by geerlingguy** [comment]
Working on this currently in #48 and in other testing

**Comment by geerlingguy** [comment]
Ugh, the mandatory setup wizard really screws up automated deployment. On CentOS, it seems that since the OS package sets up but doesn't start Jenkins immediately, we can bypass the wizard by setting `JENKINS_JAVA_OPTIONS="-Djenkins.install.runSetupWizard=false"` in the init file.

But on Ubuntu, because (I never will understand why) Deb-derivatives decide to launch the installed package immediately following installation, we can't set that option prior to the package install, therefore on first run, Jenkins builds the temporary admin password and flags the setup wizard to run.

This role then adds in that setup wizard option in the init file... but that restart is too late to affect the already-installed Jenkins instance, and the user is presented with the installation page (and commands like `jenkins-cli` don't work due to anonymous user permission restrictions).

Long-term, I'd like to make a secure-by-default installation the default. For now, I think I'll just have to accept the plugin installation won't work with 2.0 OOTB, and I'm going to leave the setup wizard in place (for now), backing out the change from #48.

Some references:

  - https://wiki.jenkins-ci.org/display/JENKINS/Jenkins+CLI
  - https://issues.jenkins-ci.org/browse/JENKINS-34035

**Comment by ronsonmaria** [comment]
thank you very much.


 Saludos,


   <http://colegio.cdsantodomingo.com>   <http://www.coiicv.org/>
<https://github.com/lronaldo/cpctelera>
    Francisco Luis Fernández Soriano
    COLEGIO DIOCESANO SANTO DOMINGO
    Calle Adolfo Claravana, 51 - 03300 ORIHUELA (Alicante)
    Telf.: 965/300 249

[image: cid:image002.gif@01CB6ACF.0C0A0970]Antes de imprimir este e-mail,
piensa bien si es necesario hacerlo. Una tonelada de papel implica la tala
de 15 árboles. Cuida el medio ambiente.

    PROTECCIÓN DE DATOS

    El COLEGIO DIOCESANO SANTO DOMINGO le informa que su dirección de
correo electrónico, así como el resto de los datos de carácter personal de
su tarjeta de visita que nos facilite, serán objeto de tratamiento
automatizado en nuestros ficheros, con la finalidad de gestionar la agenda
de contactos del Colegio, para el envío de comunicaciones profesionales y/o
personales por vía electrónica.

    Vd. podrá en cualquier momento ejercer el derecho de acceso,
rectificación, cancelación y oposición en los términos establecidos en la
Ley Orgánica 15/1999. El responsable del tratamiento es el COLEGIO
DIOCESANO SANTO DOMINGO, calle Adolfo Claravana, 51 - 03300 ORIHUELA
(Alicante).

    CONFIDENCIALIDAD

    El contenido de esta comunicación, así como el de toda la documentación
anexa, es confidencial y va dirigido únicamente al destinatario del mismo.
En el supuesto de que usted no fuera el destinatario, le solicitamos que
nos lo indique y no comunique su contenido a terceros, procediendo a su
destrucción. Gracias.

    CONFIDENCIALITY

    The content of this communication and any attached information is
confidential and exclusively for the use of the addressee. If you are not
the addressee, we ask you to notify to the sender and do not pass its
content to another person, and please be sure you destroy it. Thank you.



2016-04-27 3:00 GMT+02:00 Jeff Geerling <notifications@github.com>:

**Comment by karlmdavis** [comment]
@geerlingguy Jeff: I took a lot of inspiration from your role here when I built out my [ansible-jenkins2](https://github.com/karlmdavis/ansible-jenkins2) variant, so just thought I'd point you at my solution for the security challenges you were running into here.

I use an [init.d/adminSshKeys.groovy](https://github.com/karlmdavis/ansible-jenkins2/blob/master/templates/adminSshKeys.groovy) script to solve the security challenges you were running into here (and perhaps the other two init scripts I use also help a bit-- can't quite remember).

Feel free to steal "inspiration", yourself! Huzzah for open source!
